### ISSUE TITLE
taskMgr.step() overwrites SIGINT signal handler

### ISSUE THREAD

**Issue by cyberic99** [panda3d/panda3d#1180]
## Description
My python application uses a custom signal handler for signal 2.

After calling ``taskMgr.step()``, the original handler is not restored, and pressing ctrl+c seems to call python default signal handler instead of my custom handler.

## Steps to Reproduce
```python
#!/usr/bin/python3

import sys
import os
import signal

from direct.showbase.ShowBase import ShowBase

def my_signal_handler(sig, stack):
        print("PID %i received signal %i" % (os.getpid(), sig))
        sys.exit(0)

class MyApp(ShowBase):
    def __init__(self):
        ShowBase.__init__(self)

signal.signal(signal.SIGINT, my_signal_handler)

app = MyApp()

while True:
    taskMgr.step()
```

running this code and pressing ctrl+c produces this output:
```python
* interrupt by keyboard
Traceback (most recent call last):
  File "/home/eric/panda3d_step_signal.py", line 22, in <module>
    taskMgr.step()
  File "/home/eric/.local/lib/python3.9/site-packages/direct/task/Task.py", line 504, in step
    raise KeyboardInterrupt
KeyboardInterrupt
```
When replacing the last line with a simple ``pass``, my custom handler is called as expected.
```
PID 25887 received signal 2
```

## Environment
 * Archlinux
 * x86_64
 * Panda3D version: 1.11.0.dev2294
 * Installation method: pip
 * Python version: Python 3.9.6 (default, Jun 30 2021, 10:22:16)

**Comment by rdb** [comment]
It looks like Panda restores to the default signal handler, rather than to the handler that was assigned at the time `step()` was called.

**Comment by zachgates** [comment]
@cyberic99 Panda does a great job handling interruptions during the course of things, so integrate your code into that flow. By default, Python handles the `SIGINT` by raising `KeyboardInterrupt` (see `signal.default_int_handler`). Panda catches that exception from the `TaskManager`.

I think most people would first jump to something like this (replacing the event loop):

```python
import os

from direct.showbase.ShowBase import ShowBase


class MyApp(ShowBase):
    def run(self):
        while True:
            try:
                self.taskMgr.step()
            except KeyboardInterrupt:
                print('PID %i received SIGINT...' % os.getpid())
                break


app = MyApp()
app.run()
```

Instead, you could take advantage of the existing event loop (supplied by the `TaskManager`):

```python
import os
import sys

from direct.showbase.ShowBase import ShowBase
from direct.task import Task


class MyTaskManager(Task.TaskManager):
    def stop(self):
        super().stop()
        if sys.exc_info()[0] is KeyboardInterrupt:
            print('PID %i received SIGINT...' % os.getpid())


app = ShowBase()
app.taskMgr.__class__ = MyTaskManager
app.run()
```

**Comment by rdb** [comment]
I thought the bug was that Panda isn't restoring the signal handler at the *end* of `taskMgr.step()`, which I fixed in 2aac58b5aba8b3fef3ca6c961051d78f6458c902, but then I realised I may be misinterpreting, and you may be expecting `taskMgr.step()` to call your signal handler at the end of a frame in which Ctrl+C is pressed instead of raising KeyboardInterrupt. Is that the case?

**Comment by cyberic99** [comment]
hello @rdb , you were right, I consider the bug to be fixed with commit 2aac58b, thank you !
